openapi/openapi.yaml

```
@@ -0,0 +1,299 @@
+          text:
+            type: string
+      Metadata:
+        type: object
+        properties:
+          version:
+            type: string
+      OpportunityCategory:
+        type: object
+        properties:
+          name:
+            type: string
+            enum:
+              - 'Discretionary'
+              - 'Mandatory'
+              - 'Continuation'
+              - 'Earmark'
+              - 'Other'
+          code:
+            type: string
+            enum:
+              - 'D'
+              - 'M'
+              - 'C'
+              - 'E'
+              - 'O'
+          explanation:
+            type: string
+      Opportunity:
+        type: object
+        properties:
+          id:
+            type: string
+          number:
+            type: string
+          title:
+            type: string
+          category:
+            $ref: '#/components/schemas/OpportunityCategory'
+          required:
+            - id
+            - number
+            - title
+      ULIDType:
+        type: string
+        format: ulid
+      Grant:
+        type: object
+        properties:
+          description:
+            type: string
+          funding_instrument_types:
+            type: array
+            items:
+              $ref: '#/components/schemas/FundingInstrument'
+            minItems: 1
+          cost_sharing_or_matching_requirement:
+            type: boolean
+          CFDA_numbers:
+            type: array
+            items:
+              type: string
+            minItems: 1
+          bill:
+            type: string
+          posting_date:
+            type: string
+            format: date
+          archive_date:
+            type: string
+            format: date
+          last_updated_date:
+            type: string
+            format: date
+          eligible_applicants:
+            type: array
+            items:
+              $ref: '#/components/schemas/Applicant'
+            minItems: 1
+          additional_information:
+            $ref: '#/components/schemas/AdditionalInformation'
+          agency:
+            $ref: '#/components/schemas/Agency'
+          award:
+            $ref: '#/components/schemas/Award'
+          close_date:
+            $ref: '#/components/schemas/CloseDate'
+          funding_activity:
+            $ref: '#/components/schemas/FundingActivity'
+          grantor:
+            $ref: '#/components/schemas/GrantorContact'
+          metadata:
+            $ref: '#/components/schemas/Metadata'
+          opportunity:
+            $ref: '#/components/schemas/Opportunity'
+          revision:
+            ref: '#/components/schemas/ULIDType'
+        required:
+          - funding_instrument_types
+          - funding_activity
+          - cost_sharing_or_matching_requirement
+          - grantor
+          - posting_date
+          - eligible_applicants
+          - description
+          - CFDA_numbers
+          - opportunity
+          - revision
```

This adds the wrapper type that gets published to EventBridge. It includes a `previous` Grant (`null` when `type: create`) and a `new` Grant (`null` when `type: delete`, although we don't have a use-case for deletions at the moment). I wasn't entirely sure how to express the conditional relationship between the `type`, `new`, and `previous` fields without incurring a lot of overhead verbosity, but feel free to adjust if you think of something clever.

```suggestion
          - revision
      GrantModificationEvent:
        type: object
        properties:
          type:
            type: string
            enum:
              - create
              - update
              - delete
          versions:
            type: object
            properties:
              new:
                anyOf:
                  - type: 'null'
                  - $ref: '#/components/schemas/Grant'
              previous:
                anyOf:
                  - type: 'null'
                  - $ref: '#/components/schemas/Grant'
```

---

This looks like a solid solution to the problem you described. 👍